spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
         val joinPlan = Join(newSourceTableScan, newTargetTableScan, FullOuter, Some(cond), JoinHint.NONE)
 
-        // Construct the plan to replace the data based on the output of `MergeInto`
         val mergeParams = MergeIntoParams(
           isSourceRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_SOURCE)),
           isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_TARGET)),
-          matchedConditions = matchedActions.map(getClauseCondition),
-          matchedOutputs = matchedActions.map(actionOutput(_, target.output)),
+          matchedConditions = matchedConditions,
+          matchedOutputs = matchedOutputs,
           notMatchedConditions = notMatchedActions.map(getClauseCondition),
-          notMatchedOutputs = notMatchedActions.map(actionOutput(_, target.output)),
-          targetOutput = target.output :+ FALSE_LITERAL,
-          deleteOutput = target.output :+ TRUE_LITERAL,
+          notMatchedOutputs = notMatchedActions.map(actionOutput),
+          targetOutput = target.output,
           joinedAttributes = joinPlan.output
         )
         val mergePlan = MergeInto(mergeParams, target, joinPlan)
         val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+
         ReplaceData(target, batchWrite, mergePlan)
     }
   }
 
-  private def actionOutput(clause: MergeAction, targetOutputCols: Seq[Expression]): Seq[Expression] = {
+  private def actionOutput(clause: MergeAction): Option[Seq[Expression]] = {
     clause match {
       case u: UpdateAction =>
-        u.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(u.assignments.map(_.value))
       case _: DeleteAction =>
-        targetOutputCols :+ TRUE_LITERAL
+        None
       case i: InsertAction =>
-        i.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(i.assignments.map(_.value))
     }
   }
 
   private def getClauseCondition(clause: MergeAction): Expression = {
     clause.condition.getOrElse(TRUE_LITERAL)
   }
+
+  private def buildDynamicFilterTargetScan(
+      target: DataSourceV2Relation,
+      source: LogicalPlan,
+      cond: Expression): (MergeBuilder, LogicalPlan) = {
+    // Construct the plan to prune target based on join condition between source and target.
+    val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+    val matchingRowsPlanBuilder = (rel: DataSourceV2ScanRelation) =>
+      Join(source, rel, Inner, Some(cond), JoinHint.NONE)
+    val targetTableScan = buildDynamicFilterScanPlan(
+      target.table, target.output, mergeBuilder, cond, matchingRowsPlanBuilder)
+
+    (mergeBuilder, targetTableScan)
+  }
+
+  private def rewriteMatchedActions(
+      matchedActions: Seq[MergeAction],
+      targetOutput: Seq[Expression]): (Seq[Expression], Seq[Option[Seq[Expression]]]) = {
+    val startMatchedConditions = matchedActions.map(getClauseCondition)
+    val catchAllIndex = startMatchedConditions.indexWhere {
```

Comment:
Again for my own understanding, the fact that the `Expression` are in a `Seq` at this point in time means that they are all compounded as `and` essentially?

Reply:
These are the conditions for each action, so here each expression is independent. Evaluation will run them one-by-one until one produces `true` and then the output for that branch will be built from the corresponding outputs, `Option[Seq[Expression]]`. When that is `None`, it indicates that no output row should be produced.

Reply:
Thanks for the clarification. That makes sense.